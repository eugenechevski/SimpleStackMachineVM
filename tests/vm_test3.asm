	# $Id: vm_test3.asm,v 1.5 2024/08/29 20:26:00 leavens Exp $
	.text start
start:	STRA
	ADDI $sp, 0, 30   # stack top is 30
	SRI $sp, 1        # subtract 1 from SP (allocate a word)
	ADDI $sp, 0, 20   # top of stack is 20
	ADDI $sp, 0, 10   # stack top is 30
	BNE $sp, 1, +3    # skip if stack top and stack+1 are not equal
	CALL yes
	JMPA done
	CALL no
	JMPA done
yes:    NOTR
	PSTR $gp, 1 # print yes
	RTN
no:	NOTR
	PSTR $gp, 2 # print no
	RTN
	STRA
	EXIT 1
done:	STRA
	EXIT 0
	
	.data 1024
	WORD x = 33
	STRING[1] yeah = "Y\n"
	STRING[1] nay = "No\n"
	.stack 4096
	.end
	
