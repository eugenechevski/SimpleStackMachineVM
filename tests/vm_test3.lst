Address Instruction
     0: STRA 
     1: ADDI $sp, 0, 30
     2: SRI $sp, 1
     3: ADDI $sp, 0, 20
     4: ADDI $sp, 0, 10
     5: BNE $sp, 1, 3	# target is word address 8
     6: CALL 10	# target is word address 10
     7: JMPA 18	# target is word address 18
     8: CALL 13	# target is word address 13
     9: JMPA 18	# target is word address 18
    10: NOTR 
    11: PSTR $gp, 1
    12: RTN 
    13: NOTR 
    14: PSTR $gp, 2
    15: RTN 
    16: STRA 
    17: EXIT 1
    18: STRA 
    19: EXIT 0
    1024: 33	    1025: 2649	    1026: 683854	    1027: 0	        ...     

