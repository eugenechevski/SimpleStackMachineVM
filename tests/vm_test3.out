      PC: 0
GPR[$gp]: 1024 	GPR[$sp]: 4096 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
    1024: 33	    1025: 2649	    1026: 683854	    1027: 0	        ...     

    4096: 0	

==>      0: STRA 
      PC: 1
GPR[$gp]: 1024 	GPR[$sp]: 4096 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
    1024: 33	    1025: 2649	    1026: 683854	    1027: 0	        ...     

    4096: 0	

==>      1: ADDI $sp, 0, 30
      PC: 2
GPR[$gp]: 1024 	GPR[$sp]: 4096 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
    1024: 33	    1025: 2649	    1026: 683854	    1027: 0	        ...     

    4096: 30	

==>      2: SRI $sp, 1
      PC: 3
GPR[$gp]: 1024 	GPR[$sp]: 4095 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
    1024: 33	    1025: 2649	    1026: 683854	    1027: 0	        ...     

    4095: 0	    4096: 30	

==>      3: ADDI $sp, 0, 20
      PC: 4
GPR[$gp]: 1024 	GPR[$sp]: 4095 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
    1024: 33	    1025: 2649	    1026: 683854	    1027: 0	        ...     

    4095: 20	    4096: 30	

==>      4: ADDI $sp, 0, 10
      PC: 5
GPR[$gp]: 1024 	GPR[$sp]: 4095 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
    1024: 33	    1025: 2649	    1026: 683854	    1027: 0	        ...     

    4095: 30	    4096: 30	

==>      5: BNE $sp, 1, 3	# target is word address 8
      PC: 6
GPR[$gp]: 1024 	GPR[$sp]: 4095 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
    1024: 33	    1025: 2649	    1026: 683854	    1027: 0	        ...     

    4095: 30	    4096: 30	

==>      6: CALL 10	# target is word address 10
      PC: 10
GPR[$gp]: 1024 	GPR[$sp]: 4095 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 7    
    1024: 33	    1025: 2649	    1026: 683854	    1027: 0	        ...     

    4095: 30	    4096: 30	

==>     10: NOTR 
Y
      PC: 19
GPR[$gp]: 1024 	GPR[$sp]: 4095 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 7    
    1024: 33	    1025: 2649	    1026: 683854	    1027: 0	        ...     

    4095: 2	    4096: 30	

==>     19: EXIT 0
